### ISSUE TITLE
20+ whitespace chars in bulk insert makes index unsearchable

### ISSUE THREAD

**Issue by ifo20** [elastic/elasticsearch#32357]
**Elasticsearch version**: 6.3.2 (this doesn't occur in 6.3.1)

A bulk insert request with 20 or more consecutive whitespace chars returns a successful response but makes index unsearchable. 

The repro steps below do not require any 'initialise index' step (the issue remains whether you specify an index mapping  or use the default inferred mapping).

**Steps to reproduce**:
 1. POST request to `localhost:9200/this_index_name/_doc/_bulk`
Headers: "Content-Type" : "application/json" (have also tried "applciation/x-ndjson")
 Body:
```
{"index":{"_id":"1"}}
{"Name":"Adam"}
{"index":{"_id":"2"}}
{"Name":"Bob"}
{"index":{"_id":"3"}}
                    {"Name":"Charlie"}

```
This gives an apparently successful response:
```
{
    "took": 583,
    "errors": false,
    "items": [
        {
            "index": {
                "_index": "this_index_name",
                "_type": "_doc",
                "_id": "1",
                "_version": 1,
                "result": "created",
                "_shards": {
                    "total": 2,
                    "successful": 1,
                    "failed": 0
                },
                "_seq_no": 0,
                "_primary_term": 1,
                "status": 201
            }
        },
        {
            "index": {
                "_index": "this_index_name",
                "_type": "_doc",
                "_id": "2",
                "_version": 1,
                "result": "created",
                "_shards": {
                    "total": 2,
                    "successful": 1,
                    "failed": 0
                },
                "_seq_no": 0,
                "_primary_term": 1,
                "status": 201
            }
        },
        {
            "index": {
                "_index": "this_index_name",
                "_type": "_doc",
                "_id": "3",
                "_version": 1,
                "result": "created",
                "_shards": {
                    "total": 2,
                    "successful": 1,
                    "failed": 0
                },
                "_seq_no": 0,
                "_primary_term": 1,
                "status": 201
            }
        }
    ]
}

```
2. Try and search the index : `localhost:9200/this_index_name/_search`
Get the following error response:
```
{
    "error": {
        "root_cause": [
        {
            "type": "illegal_argument_exception",
            "reason": "Can't write raw bytes whose xcontent-type can't be guessed"
        }
        ],
    "type": "illegal_argument_exception",
    "reason": "Can't write raw bytes whose xcontent-type can't be guessed"
},
    "status": 400
}

```

There don't appear to be any relevant logs but here they are:
```[2018-07-25T10:53:51,278][INFO ][o.e.n.Node               ] [] initializing ...
[2018-07-25T10:53:51,348][INFO ][o.e.e.NodeEnvironment    ] [9kDaYeG] using [1] data paths, mounts [[OS (C:)]], net usable_space [84.6gb], net total_space [232.3gb], types [NTFS]
[2018-07-25T10:53:51,348][INFO ][o.e.e.NodeEnvironment    ] [9kDaYeG] heap size [989.8mb], compressed ordinary object pointers [unknown]
[2018-07-25T10:53:51,351][INFO ][o.e.n.Node               ] [9kDaYeG] node name derived from node ID [9kDaYeGGTL-IO9TLc4IfUw]; set [node.name] to override
[2018-07-25T10:53:51,352][INFO ][o.e.n.Node               ] [9kDaYeG] version[6.3.2], pid[19936], build[default/zip/053779d/2018-07-20T05:20:23.451332Z], OS[Windows 10/10.0/x86], JVM[Oracle Corporation/Java HotSpot(TM) Client VM/1.8.0_171/25.171-b11]
[2018-07-25T10:53:51,352][INFO ][o.e.n.Node               ] [9kDaYeG] JVM arguments [-Xms1g, -Xmx1g, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Djava.io.tmpdir=C:\Users\IOlliver\AppData\Local\Temp\elasticsearch, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=data, -XX:ErrorFile=logs/hs_err_pid%p.log, -XX:+PrintGCDetails, -XX:+PrintGCDateStamps, -XX:+PrintTenuringDistribution, -XX:+PrintGCApplicationStoppedTime, -Xloggc:logs/gc.log, -XX:+UseGCLogFileRotation, -XX:NumberOfGCLogFiles=32, -XX:GCLogFileSize=64m, -Delasticsearch, -Des.path.home=C:\Users\IOlliver\Downloads\elasticsearch-6.3.2\elasticsearch-6.3.2, -Des.path.conf=C:\Users\IOlliver\Downloads\elasticsearch-6.3.2\elasticsearch-6.3.2\config, -Des.distribution.flavor=default, -Des.distribution.type=zip]
[2018-07-25T10:53:53,933][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [aggs-matrix-stats]
[2018-07-25T10:53:53,933][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [analysis-common]
[2018-07-25T10:53:53,933][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [ingest-common]
[2018-07-25T10:53:53,933][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [lang-expression]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [lang-mustache]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [lang-painless]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [mapper-extras]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [parent-join]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [percolator]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [rank-eval]
[2018-07-25T10:53:53,934][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [reindex]
[2018-07-25T10:53:53,935][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [repository-url]
[2018-07-25T10:53:53,935][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [transport-netty4]
[2018-07-25T10:53:53,935][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [tribe]
[2018-07-25T10:53:53,935][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-core]
[2018-07-25T10:53:53,935][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-deprecation]
[2018-07-25T10:53:53,935][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-graph]
[2018-07-25T10:53:53,936][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-logstash]
[2018-07-25T10:53:53,936][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-ml]
[2018-07-25T10:53:53,937][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-monitoring]
[2018-07-25T10:53:53,937][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-rollup]
[2018-07-25T10:53:53,938][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-security]
[2018-07-25T10:53:53,939][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-sql]
[2018-07-25T10:53:53,939][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-upgrade]
[2018-07-25T10:53:53,940][INFO ][o.e.p.PluginsService     ] [9kDaYeG] loaded module [x-pack-watcher]
[2018-07-25T10:53:53,941][INFO ][o.e.p.PluginsService     ] [9kDaYeG] no plugins loaded
[2018-07-25T10:53:58,884][INFO ][o.e.x.s.a.s.FileRolesStore] [9kDaYeG] parsed [0] roles from file [C:\Users\IOlliver\Downloads\elasticsearch-6.3.2\elasticsearch-6.3.2\config\roles.yml]
[2018-07-25T10:53:59,551][DEBUG][o.e.a.ActionModule       ] Using REST wrapper from plugin org.elasticsearch.xpack.security.Security
[2018-07-25T10:53:59,743][INFO ][o.e.d.DiscoveryModule    ] [9kDaYeG] using discovery type [zen]
[2018-07-25T10:54:00,359][INFO ][o.e.n.Node               ] [9kDaYeG] initialized
[2018-07-25T10:54:00,360][INFO ][o.e.n.Node               ] [9kDaYeG] starting ...
[2018-07-25T10:54:00,726][INFO ][o.e.t.TransportService   ] [9kDaYeG] publish_address {127.0.0.1:9300}, bound_addresses {127.0.0.1:9300}, {[::1]:9300}
[2018-07-25T10:54:00,746][WARN ][o.e.b.BootstrapChecks    ] [9kDaYeG] JVM is using the client VM [Java HotSpot(TM) Client VM] but should be using a server VM for the best performance
[2018-07-25T10:54:03,791][INFO ][o.e.c.s.MasterService    ] [9kDaYeG] zen-disco-elected-as-master ([0] nodes joined)[, ], reason: new_master {9kDaYeG}{9kDaYeGGTL-IO9TLc4IfUw}{lupi21suR0GclAZR-fjSZQ}{127.0.0.1}{127.0.0.1:9300}{xpack.installed=true}
[2018-07-25T10:54:03,798][INFO ][o.e.c.s.ClusterApplierService] [9kDaYeG] new_master {9kDaYeG}{9kDaYeGGTL-IO9TLc4IfUw}{lupi21suR0GclAZR-fjSZQ}{127.0.0.1}{127.0.0.1:9300}{xpack.installed=true}, reason: apply cluster state (from master [master {9kDaYeG}{9kDaYeGGTL-IO9TLc4IfUw}{lupi21suR0GclAZR-fjSZQ}{127.0.0.1}{127.0.0.1:9300}{xpack.installed=true} committed version [1] source [zen-disco-elected-as-master ([0] nodes joined)[, ]]])
[2018-07-25T10:54:03,842][WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [9kDaYeG] Failed to clear cache for realms [[]]
[2018-07-25T10:54:03,883][INFO ][o.e.g.GatewayService     ] [9kDaYeG] recovered [0] indices into cluster_state
[2018-07-25T10:54:03,951][INFO ][o.e.x.s.t.n.SecurityNetty4HttpServerTransport] [9kDaYeG] publish_address {127.0.0.1:9200}, bound_addresses {127.0.0.1:9200}, {[::1]:9200}
[2018-07-25T10:54:03,952][INFO ][o.e.n.Node               ] [9kDaYeG] started
[2018-07-25T10:54:03,999][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.triggered_watches] for index patterns [.triggered_watches*]
[2018-07-25T10:54:04,042][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.watch-history-7] for index patterns [.watcher-history-7*]
[2018-07-25T10:54:04,067][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.watches] for index patterns [.watches*]
[2018-07-25T10:54:04,103][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.monitoring-logstash] for index patterns [.monitoring-logstash-6-*]
[2018-07-25T10:54:04,154][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.monitoring-es] for index patterns [.monitoring-es-6-*]
[2018-07-25T10:54:04,178][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.monitoring-beats] for index patterns [.monitoring-beats-6-*]
[2018-07-25T10:54:04,201][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.monitoring-alerts] for index patterns [.monitoring-alerts-6]
[2018-07-25T10:54:04,225][INFO ][o.e.c.m.MetaDataIndexTemplateService] [9kDaYeG] adding template [.monitoring-kibana] for index patterns [.monitoring-kibana-6-*]
[2018-07-25T10:54:04,267][INFO ][o.e.l.LicenseService     ] [9kDaYeG] license [48f1088e-a83e-4126-9751-e5f75956d7e7] mode [basic] - valid
[2018-07-25T10:54:36,424][INFO ][o.e.c.m.MetaDataCreateIndexService] [9kDaYeG] [autocomplete_tenant_test] creating index, cause [auto(bulk api)], templates [], shards [5]/[1], mappings []
[2018-07-25T10:54:36,905][INFO ][o.e.c.m.MetaDataMappingService] [9kDaYeG] [autocomplete_tenant_test/Pm8y5s4hQLCPfMFldTxQ3g] create_mapping [_doc]```

**Comment by andrershov** [comment]
The issue is reproducible on 6.x and on master. I can not confirm that the issue was introduced in 6.3.1. The issue is still reproducible in 6.3.1 and 6.3.0.
The exception is happening in JsonXContentGenerator.writeRawField, while trying to guess "Content-Type" for _source field. If Content-Type could not be guessed, IllegalArgumentException is thrown. XContentType factory is using GUESS_HEADER_LENGTH (==20) first characters to guess the type of the field. Since they are all whitespaces - null is returned.
This issue is reproducible with non-bulk API as well - as long as _source starts with 20 or whitespaces.

**Comment by bleskes** [comment]
@andrershov I need some more details to be fully sure (you mention writeRawField, but I don't yet see how this connects to writing), but it seems this is something specific to how the bulk api does things. since 6.0 content type is required with each call to our APIs, and it's properly supplied in the request in this issue. From this, I think this is a bug that we even end up trying to detect what the content type is. 

You mention:
> This issue is reproducible with non-bulk API as well - as long as _source starts with 20 or more whitespaces.

which `_source` do you refer to? can you give a complete example?

PS I'm going to re-label w.r.t the Bulk API until we clarify it's a problem with a more generic infra.

**Comment by jasontedor** [comment]
@bleskes The problem is on the search response, not the bulk request. We don’t store the source content-type and content-type detection is failing then to write the search response. The content-type detection fails because we only look at the first twenty bytes of source to guess the content-type. I don’t think we should modify user source ever. We can consider being stricter about the format of a bulk request, it depends what we document here whether or not we can get away with what would otherwise be a breaking change. We can also consider modifying content-type detection to strip leading whitespace. The latter feels like a core issue to me.

**Comment by bleskes** [comment]
@jasontedor thanks. @andrershov and I discussed it over another channel and he clarified it. W.r.t improve content type detection - I'm +1 on that , though it's a bit tricky to do imo. `xContentType(InputStream si)` shouldn't change the stream, ideally and if it does it will change the response - which kind of defeats the point of not messing up with the `_source`. 

I've asked @andrershov to also look at what it will take to validating incoming indexing ops to make sure we can guess the content type correctly (and ideally it being equal to what we're being told). I think that's better (if it works out) until we properly resolve the "what content type do we have stored?" discussion.

**Comment by jasontedor** [comment]
> W.r.t improve content type detection - I'm +1 on that , though it's a bit tricky to do imo. `xContentType(InputStream si)` shouldn't change the stream, ideally and if it does it will change the response - which kind of defeats the point of not messing up with the `_source`. 

I don't think so. I think we can do a little dance with mark/reset. 💃

**Comment by jasontedor** [comment]
I opened #32632.

**Comment by jasontedor** [comment]
With the change in #32632 this issue no longer reproduces:

```
13:38:10 [jason@totoro:~] $ cat bulk.txt                                                                                                                                                                                                                                                                          
{"index":{"_id":"1"}}                                                                                                                                                                                                                                                                                             
{"Name":"Adam"}                                                                                                                                                                                                                                                                                                   
{"index":{"_id":"2"}}                                                                                                                                                                                                                                                                                             
{"Name":"Bob"}                                                                                                                                                                                                                                                                                                    
{"index":{"_id":"3"}}                                                                                                                                                                                                                                                                                             
                    {"Name":"Charlie"}                                                                                                                                                                                                                                                                            
13:38:16 [jason@totoro:~] $ curl -XDELETE localhost:9200/_all?pretty=true                                                                                                                                                                                                                                         
{                                                                                                                                                                                                                                                                                                                 
  "acknowledged" : true                                                                                                                                                                                                                                                                                           
}                                                                                                                                                                                                                                                                                                                 
13:38:21 [jason@totoro:~] $ curl -XPOST -H "Content-Type: application/json" localhost:9200/i/_doc/_bulk?pretty=true --data-binary @bulk.txt                                                                                                                                                                       
{                                                                                                                                                                                                                                                                                                                 
  "took" : 174,         
  "errors" : false,
  "items" : [
    {  
      "index" : {      
        "_index" : "i",  
        "_type" : "_doc",
        "_id" : "1",   
        "_version" : 1,
        "result" : "created",
        "_shards" : {
          "total" : 2,
          "successful" : 1,
          "failed" : 0
        },
        "_seq_no" : 0,     
        "_primary_term" : 1,
        "status" : 201
      }
    },
    {
      "index" : {
        "_index" : "i",
        "_type" : "_doc",
        "_id" : "2",
        "_version" : 1,
        "result" : "created",
        "_shards" : {
          "total" : 2,
          "successful" : 1,
          "failed" : 0
        },
        "_seq_no" : 1,
        "_primary_term" : 1,
        "status" : 201
      }
    },
    {
      "index" : {
        "_index" : "i",
        "_type" : "_doc",
        "_id" : "3",
        "_version" : 1,
        "result" : "created",
        "_shards" : {
          "total" : 2,
          "successful" : 1,
          "failed" : 0
        },
        "_seq_no" : 2,
        "_primary_term" : 1,
        "status" : 201
      }
    }
  ]
}
13:38:29 [jason@totoro:~] $ curl -XGET localhost:9200/_search?pretty=true
{
  "took" : 6,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 3,
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "i",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : 1.0,
        "_source" : {
          "Name" : "Adam"
        }
      },
      {
        "_index" : "i",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 1.0,
        "_source" : {
          "Name" : "Bob"
        }
      },
      {
        "_index" : "i",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : 1.0,
        "_source" : {
          "Name" : "Charlie"
        }
      }
    ]
  }
}
```

**Comment by jasontedor** [comment]
Closed by #32632
